include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -262,34 +262,43 @@ class operational_domain_impl
             params{ps},
             stats{st},
             output_bdl_pairs{detect_bdl_pairs<Lyt>(layout, sidb_technology::cell_type::OUTPUT, params.bdl_params)},
-            x_indices(num_x_steps()),  // pre-allocate the x dimension indices
-            y_indices(num_y_steps()),  // pre-allocate the y dimension indices
-            x_values(num_x_steps()),   // pre-allocate the x dimension values
-            y_values(num_y_steps())    // pre-allocate the y dimension values
+            x_indices(num_x_steps() + 1),  // pre-allocate the x dimension indices
+            y_indices(num_y_steps() + 1)   // pre-allocate the y dimension indices
     {
+        x_values.reserve(num_x_steps() + 1);
+        y_values.reserve(num_y_steps() + 1);
+
         op_domain.x_dimension = params.x_dimension;
         op_domain.y_dimension = params.y_dimension;
 
         std::iota(x_indices.begin(), x_indices.end(), 0ul);
         std::iota(y_indices.begin(), y_indices.end(), 0ul);
 
+        if ((params.x_min + (x_indices.size() - 1) * params.x_step) - params.x_max >
+            physical_constants::POP_STABILITY_ERR)
+        {
+            x_indices.pop_back();
+        }
+        if (((params.y_min + (y_indices.size() - 1) * params.y_step) - params.y_max) >
+            physical_constants::POP_STABILITY_ERR)
+        {
+            y_indices.pop_back();
+        }
+
         // generate the x dimension values
-        std::generate(x_values.begin(), x_values.end(),
-                      [x = 0, this]() mutable
-                      {
-                          const double x_val = params.x_min + x * params.x_step;
-                          ++x;
-                          return x_val;
-                      });
+        auto x_val = params.x_min;
+        for (size_t y = 0; y < x_indices.size() && x_val < params.x_max + physical_constants::POP_STABILITY_ERR; ++y)
+        {
+            x_values.push_back(x_val);
+            x_val += params.x_step;
+        }
```

Comment:
Is `x_val` of a floating-point type (I assume `double`)? These should never be used as loop counters due to accumulating imprecision.